### PR TITLE
Add norms about testing

### PR DESCRIPTION
**Pull request by NickHeiner**
Related: #2333.

### CODE REVIEW

File: docs/testing-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+As much as possible, business logic should exist in pure functions, which is a function whose outputs are only reflective of its inputs, and which does not cause side effects. These functions are easy to unit test.
+
+On the frontend, Redux reducers should have unit tests. Components can also have unit tests via Enzyme. 
+
+The backend should be tested with both unit tests and request-level tests.
+
+The backend and frontend are tested together with Capybara tests. Because Capybara is slow and it's hard to write reliable tests, it should be used as more of a smoke test than an exhaustive proof of correctness. Try one or two permutations in Capybara, and verify all the edge cases in unit tests.
+
+Before deployment, Artem's full stack integration tests will also run.
```
Comment (cmgiven) [inline docs/testing-guidelines.md]: Even though he's written most of them, we probably shouldn't describe any single person as the sole owner of tests, even an integration testing hero like Artem.

File: docs/testing-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+As much as possible, business logic should exist in pure functions, which is a function whose outputs are only reflective of its inputs, and which does not cause side effects. These functions are easy to unit test.
+
+On the frontend, Redux reducers should have unit tests. Components can also have unit tests via Enzyme. 
+
+The backend should be tested with both unit tests and request-level tests.
+
+The backend and frontend are tested together with Capybara tests. Because Capybara is slow and it's hard to write reliable tests, it should be used as more of a smoke test than an exhaustive proof of correctness. Try one or two permutations in Capybara, and verify all the edge cases in unit tests.
+
+Before deployment, QA's full stack integration tests will also run.
```
Comment (kierachell) [inline docs/testing-guidelines.md]: Might as well link to them at: https://github.com/department-of-veterans-affairs/appeals-qa/blob/master/smokeTests/spec.md

File: docs/testing-guidelines.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+Before deployment, QA's full stack integration tests will also run. 
+
+## Never Sleep
+No test should contain a `sleep` or `await pause()` where we are guessing how long a timed operation will take. This is a recipe for flakey tests. If we guess too low a time, the test will fail. If we guess too high a time, the test will be unnecessarily slow.
+
+Instead, do a "spinning assert", where we continually check for the condition we're waiting for. For example:
+
+```js
+while (true) {
```
Comment (mdbenjam) [inline docs/testing-guidelines.md]: This won't work in javascript since it's single threaded. You need to use something like a setTimeout, to give other threads a chance to run.

File: docs/testing-guidelines.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+On the frontend, Redux reducers should have unit tests. Components can also have unit tests via Enzyme. 
+
+The backend should be tested with both unit tests and request-level tests.
+
+The backend and frontend are tested together with Capybara tests. Because Capybara is slow and it's hard to write reliable tests, it should be used as more of a smoke test than an exhaustive proof of correctness. Try one or two permutations in Capybara, and verify all the edge cases in unit tests.
+
+Before deployment, QA's full stack integration tests will also run. 
+
+## Never Sleep
+No test should contain a `sleep` or `await pause()` where we are guessing how long a timed operation will take. This is a recipe for flakey tests. If we guess too low a time, the test will fail. If we guess too high a time, the test will be unnecessarily slow.
```
Comment (mdbenjam) [inline docs/testing-guidelines.md]: I think await pause isn't as bad as sleeping. Agree that polling is better, but you need a mechanism for allowing other threads to run, since javascript is single threaded.

File: docs/testing-guidelines.md (hunk 3 above)
Comment (NickHeiner) [inline docs/testing-guidelines.md]: Good point. Changed this to pseudocode. A fully-working JS example would be more complex than what I want at this point in the doc.

File: docs/testing-guidelines.md (hunk 4 above)
Comment (NickHeiner) [inline docs/testing-guidelines.md]: `await pause()` is sleeping.
